Ensure satellite resources are published when building in VS

pranavkm opened the pull request.

pranavkm committed:
Ensure satellite resources are published when building in VS

SteveSandersonMS commented on src/Components/WebAssembly/Build/src/targets/Blazor.MonoRuntime.targets:
```diff
@@ -140,6 +147,41 @@
+    <BlazorWriteSatelliteAssemblyFile
+      SatelliteAssembly="@(_BlazorWriteSatelliteAssembly)"
+      WriteFile="$(_BlazorSatelliteAssemblyStashFile)"
+      Condition="'$(BuildInsideVisualStudio)' == 'true' AND '$(BuildingProject)' == 'true' AND '@(_BlazorWriteSatelliteAssembly->Count())' != '0'" />
```
Are the conditions `'$(BuildInsideVisualStudio)' == 'true'` and `'@(_BlazorWriteSatelliteAssembly->Count())' != '0'` just a way of trying to minimize the risk of negatively impacting other scenarios?

Because it seems like we should really be doing this even for command-line builds, in case someone later does a publish with `--no-build`, right? And the `_BlazorWriteSatelliteAssembly.Count()` == 0 scenario shouldn't be harmful (we'd just write/read an empty list).

SteveSandersonMS commented on src/Components/WebAssembly/Build/src/targets/Blazor.MonoRuntime.targets:
```diff
@@ -75,6 +75,13 @@
 
+  <Import
+    Project="$(_BlazorSatelliteAssemblyStashFile)"
+    Condition="Exists('$(_BlazorSatelliteAssemblyStashFile)')" />
```
I haven't yet understood this part. Why do we `<Import>` the stash file rather than use `<BlazorReadSatelliteAssemblyFile>` to read it?

The `BlazorReadSatelliteAssemblyFile` part I'm pretty comfortable. That's the same kind of "store a manifest file" pattern we use in loads of places. But the technique of "write out an actual `.targets` file and then import it" is quite novel.

Presumably this is essential for some reason otherwise you wouldn't be doing it, but if so, could you clarify why it's not sufficient to write the data in whatever custom format we want (e.g., `.xml` file) and then use `BlazorReadSatelliteAssemblyFile` to read it back and generate a corresponding ItemGroup?

pranavkm committed:
Add more tests

pranavkm commented:
🆙 📅 

* Remove the check for VS
* Added some build incrementalism tests
* Fixed up a scenario to clean up the cache file once resx files are removed.

pranavkm committed:
Include pwa check

SteveSandersonMS commented on src/Components/WebAssembly/Build/test/BuildIntegrationTests/PublishIntegrationTest.cs:
```diff
@@ -433,6 +432,109 @@ public async Task Publish_HostedApp_WithNoBuild_Works()
+            VerifyServiceWorkerFiles(result, blazorPublishDirectory,
+                serviceWorkerPath: Path.Combine("serviceworkers", "my-service-worker.js"),
+                serviceWorkerContent: "// This is the production service worker",
+                assetsManifestPath: "custom-service-worker-assets.js");
```
How come this template produces a service worker? Looks like it’s just the hosted template, not a PWA.

pranavkm replied:
The standalone app has a service worker, no? https://github.com/dotnet/aspnetcore/blob/blazor-wasm/src/Components/WebAssembly/Build/testassets/standalone/standalone.csproj#L7

SteveSandersonMS reviewed (approved):
Fantastic!
